.github/workflows/publish.yaml

```diff
@@ -0,0 +1,94 @@
+name: Publish
+
+on:
+  push:
+    branches:
+      - master
+      - main
+      - track/**
+  pull_request:
+    branches:
+      - master
+      - main
+      - track/**
+
+jobs:
+  lib-check:
+    name: Check libraries
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Check libs
+        uses: canonical/charming-actions/check-libraries@1.0.3
+        with:
+          credentials: "${{ secrets.CHARMCRAFT_CREDENTIALS }}"
+          github-token: "${{ secrets.GITHUB_TOKEN }}"
+
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Install dependencies
+        run: python3 -m pip install tox
+      - name: Run linters
+        run: tox -e lint
+
+  unit-test:
+    name: Unit tests
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Install dependencies
+        run: python -m pip install tox
+      - name: Run tests
+        run: tox -e unit
+
+  integration-test-microk8s:
+    name: Integration tests (microk8s)
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Setup operator environment
+        uses: charmed-kubernetes/actions-operator@main
+        with:
+          provider: microk8s
+      - name: Run integration tests
+        # set a predictable model name so it can be consumed by charm-logdump-action
+        run: tox -e integration -- --model testing
+      - name: Dump logs
+        uses: canonical/charm-logdump-action@main
+        if: failure()
+        with:
+          # TEMPLATE-TODO: Replace the application name
+          app: operator-template
+          model: testing
```

Comment:
This duplicates everything from `ci.yaml`, so now maintainers need to write their testing twice (once in `ci.yaml`, once here).  I think I know what you're trying to do and I think we can achieve the goal without repeating the testing.

Reply:
Is the goal here to achieve:
* only publish to charmhub if CI passes (or my personal preference, only publish to `latest/edge` if CI passes, but always publish to `latest/edge/branchname` even if CI fails)
?

If yes, I think we can achieve this through a bit different methods without needing to repeat the tests

Reply:
I think there's at least two ways to achieve this.  Would appreciate other's feedback!

### Using workflow dispatch to reuse a workflow
See [reusing workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) for more detail, but effectively we can call an entire workflow similar to how we can call an action.  So we could do one of:
1. our `test.yaml` will call `publish.yaml` in cases where we should publish
2. our `publish.yaml` will call `test.yaml` before it runs

Either way I think it works.  It'll take a bit more effort to get the "publish to `latest/edge` IFF passing `test.yaml`, but always publish to `latest/edge/branch`", but there should be a way.

### Publishing to branch on `pull_request` events, publishing to `*/edge` on `push` events on selected branches.  

We could separate our publishing into two things:
```yaml
# publish-to-branch.yaml
on:
  pull_request:

#... publish
```

```yaml
# publish-to-edge.yaml
on:
  push:
    branches:
      - main
      - master
      - track/*

# workflow dispatch or otherwise call `test.yaml` to get tests

# publish, with `needs: above-tests`
```

I think this achieves all goals.

#### Variation on above using repo settings to protect against bad publishing

Instead of `publish-to-edge.yaml` above invoking a workflow dispatch, we could instead use repo settings to achieve the same thing.  `Settings->branchs->Branch protection rules` let us:
* `require "status checks" to pass before merging` anything into `main`/`track/*`:  this lets us require that all tests in `test.yaml` have passed before merging into a branch that gets published
* `require branches to be up to date before merging`: this lets us avoid "stale" mergers (where the CI would have failed because main has changed recently)

Reply:
From discussions with @sed-i, this is a possible way to implement the "using workflow dispatch" option.  We have workflow files of:

* `tests.yaml` <-- a workflow_dispatchable workflow for tests (lint, unit, ...)
* `publish.yaml` <-- a workflow_dispatchable workflow for publishing
* `on-pr.yaml` <-- on any pull_request event, this runs a workflow_dispatch for tests.yaml and publish.yaml (to branch), where publish runs no matter what
* `on-push.yaml` <-- on any push, this runs tests.yaml and then if tests passpublish.yaml (to edge)